include/ensmallen_bits/adamr/adamr.hpp

```diff
@@ -0,0 +1,170 @@
+/**
+ * @file adamr.hpp
+ * @author Niteya Shah
+ * Declaration of the AdamR optimizer.
+ *
+ * ensmallen is free software; you may redistribute it and/or modify it under
+ * the terms of the 3-clause BSD license.  You should have received a copy of
+ * the 3-clause BSD license along with ensmallen.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef ENSMALLEN_ADAM_ADAMR_HPP
+#define ENSMALLEN_ADAM_ADAMR_HPP
+
+#include <ensmallen_bits/adam/adam_update.hpp>
+#include <ensmallen_bits/sgdr/cyclical_decay.hpp>
+#include <ensmallen_bits/sgd/sgd.hpp>
+#include <ensmallen_bits/adam/adamw_update.hpp>
+
+namespace ens {
+/**
+ * This class is based on the Adam Update where the Optimiser
+ * simulates a new warm-started run/restart once a number of epochs are
+ * performed.
+ *
+ * @code
+ * @article{
+ *   title   = {Decoupled Weight Decay Regularization},
+ *   author  = {{Ilya}, L. and {Frank}, H.},
+ *   journal = {ArXiv e-prints},
+ *   url     = {https://arxiv.org/pdf/1711.05101.pdf}
+ *   year    = {2019}
+ * }
+ * @endcode
+ *
+ * AdamR can optimize differentiable separable functions.  For more details, see
+ * the documentation on function types included with this distribution or on the
+ * ensmallen website.
+ * @tparam UpdateRule Adam optimizer update rule to be used.
+ * @tparam DecayPolicy The StepSize decay policy to be used by default
+ * CyclicalDecay.
+*/
+template<typename UpdateRule = AdamUpdate,typename DecayPolicyType = CyclicalDecay>
+class AdamRType
+{
+ public:
+  /**
+   * Construct the AdamR optimizer with the given function and parameters. The
+   * defaults here are not necessarily good for the given problem, so it is
+   * suggested that the values used be tailored to the task at hand.  The
+   * maximum number of iterations refers to the maximum number of points that
+   * are processed (i.e., one iteration equals one point; one iteration does not
+   * equal one pass over the dataset).
+   *
+   * @param stepSize Maximum and initial step size for each batch of warm
+                     restart.
+   * @param epochRestart Restart Rate for Warm Restarts
+   * @param multFactor Multiplier for epochRestart
+   * @param stepSizeMin Minimum and final step size for each batch of warm
+                        restart (Use DecayPolicy() method with CyclicalDecay).
```

Comment:
`stepSizeMin` doesn't appear in the list below.  In addition, I would suggest that perhaps we put this right next to the `stepSize` parameter.